### PR TITLE
bench: Add bench for mvcctxn

### PR DESCRIPTION
**Pull request by MyonKeminta**
Implemented bench for single-row get, set and delete.

### CODE REVIEW

File: benches/bin/mvcctxn.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+use tikv::storage::mvcc::MvccTxn;
+use kvproto::kvrpcpb::{Context, IsolationLevel};
+
+use super::print_result;
+use test::BenchSamples;
+
+use rand::Rng;
+
+
+// TODO: Make this thread safe
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: seem you can use atomic

File: benches/bin/mvcctxn.rs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    do_write(engine, txn.into_modifies());
+}
+
+fn prepare_test_engine(
+    versions: usize,
+    value_len: usize,
+    keys: &[Vec<u8>],
+    values: &mut Vec<Vec<u8>>,
+) -> Box<Engine> {
+    let engine = new_local_engine(TEMP_DIR, ALL_CFS).unwrap();
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: I think we should the default TiKV configuration here.

**Comment by siddontang** [comment]
Great work

Can you show some benchmark results?

File: benches/bin/mvcctxn.rs (hunk 2 above)
Comment (MyonKeminta) [reply benches/bin/mvcctxn.rs]: I saw the implementation  of  `new_local_engine` function, and I think it does used the default configurations.

File: benches/bin/mvcctxn.rs (hunk 1 above)
Comment (MyonKeminta) [reply benches/bin/mvcctxn.rs]: I'll patch it in the next commit

**Comment by MyonKeminta** [comment]
On my not-so-powerful laptop, it gives following result
```
benching mvcctxn get    rows:1000 versions:5 value len:128      ...     96,109 ns/iter (+/- 2,050)
benching mvcctxn set    rows:1000 versions:5 value len:128      ...    384,410 ns/iter (+/- 22,359)
benching mvcctxn delete rows:1000 versions:5 value len:128      ...    513,713 ns/iter (+/- 45,110)
benching mvcctxn get    rows:10000 versions:5 value len:128     ...    190,824 ns/iter (+/- 14,160)
benching mvcctxn set    rows:10000 versions:5 value len:128     ...    393,851 ns/iter (+/- 29,988)
benching mvcctxn delete rows:10000 versions:5 value len:128     ...    925,869 ns/iter (+/- 656,527)
benching mvcctxn get    rows:100000 versions:5 value len:128    ...    302,308 ns/iter (+/- 10,192)
benching mvcctxn set    rows:100000 versions:5 value len:128    ...    759,485 ns/iter (+/- 106,805)
benching mvcctxn delete rows:100000 versions:5 value len:128    ...    674,219 ns/iter (+/- 95,984)
benching mvcctxn get    rows:10000 versions:1 value len:128     ...    112,877 ns/iter (+/- 5,179)
benching mvcctxn set    rows:10000 versions:1 value len:128     ...    569,377 ns/iter (+/- 205,909)
benching mvcctxn delete rows:10000 versions:1 value len:128     ...    516,627 ns/iter (+/- 30,999)
benching mvcctxn get    rows:10000 versions:5 value len:128     ...    191,584 ns/iter (+/- 12,550)
benching mvcctxn set    rows:10000 versions:5 value len:128     ...    841,070 ns/iter (+/- 240,264)
benching mvcctxn delete rows:10000 versions:5 value len:128     ...    348,630 ns/iter (+/- 46,792)
benching mvcctxn get    rows:10000 versions:20 value len:128    ...    200,294 ns/iter (+/- 24,201)
benching mvcctxn set    rows:10000 versions:20 value len:128    ...    413,164 ns/iter (+/- 31,058)
benching mvcctxn delete rows:10000 versions:20 value len:128    ...    559,177 ns/iter (+/- 64,886)
benching mvcctxn get    rows:10000 versions:5 value len:32      ...    216,139 ns/iter (+/- 21,240)
benching mvcctxn set    rows:10000 versions:5 value len:32      ...    740,535 ns/iter (+/- 143,281)
benching mvcctxn delete rows:10000 versions:5 value len:32      ...    354,416 ns/iter (+/- 21,282)
benching mvcctxn get    rows:10000 versions:5 value len:128     ...    119,427 ns/iter (+/- 4,316)
benching mvcctxn set    rows:10000 versions:5 value len:128     ...    592,193 ns/iter (+/- 62,216)
benching mvcctxn delete rows:10000 versions:5 value len:128     ...    745,521 ns/iter (+/- 320,384)
benching mvcctxn get    rows:10000 versions:5 value len:1024    ...    191,774 ns/iter (+/- 95,467)
benching mvcctxn set    rows:10000 versions:5 value len:1024    ...  1,025,160 ns/iter (+/- 125,540)
benching mvcctxn delete rows:10000 versions:5 value len:1024    ...    351,210 ns/iter (+/- 18,715)
```

**Comment by siddontang** [comment]
PTAL @wentaoxu

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987 @disksing

File: benches/bin/mvcctxn.rs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        rng.fill_bytes(&mut value);
+
+        let key = &keys[rng.gen_range(0, keys.len())];
+
+        //let check_value = value.clone();
+
+        prewrite(engine, &[Mutation::Put((Key::from_raw(key), value))], key, start_ts);
+        commit(engine, &[Key::from_raw(key)], start_ts, commit_ts)
+
+        // Debug code
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: please remove the unnecessary codes.

**Comment by wentaoxu** [comment]
LGTM

File: benches/bin/mvcctxn.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+#[inline]
+fn next_ts() -> u64 {
+    static CURRENT: AtomicU64 = ATOMIC_U64_INIT;
+    CURRENT.fetch_add(1, Ordering::SeqCst)
+}
+
+#[inline]
+fn to_bytes(value: u64) -> Vec<u8> {
+    let mut result = Vec::new();
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: why not using byteorder library directly?

File: benches/bin/mvcctxn.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    // big endian
+    for shift in (0..8).rev() {
+        result.push(((value >> shift) & 0xff) as u8);
+    }
+    result
+}
+
+fn create_row_key(tableid: u64, rowid: u64) -> Vec<u8> {
+    let mut result = to_bytes(tableid);
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: should pre-allocate the buffer to avoid reallocation.

File: benches/bin/mvcctxn.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    result.extend(b"_r".iter());
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: you can use `extend_from_slice(b"_r")` instead

File: benches/bin/mvcctxn.rs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+fn to_bytes(value: u64) -> Vec<u8> {
+    let mut result = Vec::new();
+    // big endian
+    for shift in (0..8).rev() {
+        result.push(((value >> shift) & 0xff) as u8);
+    }
+    result
+}
+
+fn create_row_key(tableid: u64, rowid: u64) -> Vec<u8> {
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: please refer https://github.com/pingcap/tikv/blob/master/src/raftstore/store/keys.rs#L68 to see how we optimize the key creation.

File: benches/bin/mvcctxn.rs [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    let mut result = to_bytes(tableid);
+    result.extend(b"_r".iter());
+    result.extend(to_bytes(rowid).iter());
+    result
+}
+
+fn generate_keys(count: usize) -> Vec<Vec<u8>> {
+    let mut result = Vec::new();
```
Comment (siddontang) [inline benches/bin/mvcctxn.rs]: create with capacity.

File: benches/bin/mvcctxn.rs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+fn create_row_key(tableid: i64, rowid: i64) -> Vec<u8> {
+    let mut handle = Vec::with_capacity(8);
+    handle.encode_i64(rowid).unwrap();
+    encode_row_key(tableid, &handle)
+}
+
+fn generate_keys(count: usize) -> Vec<Vec<u8>> {
+    let mut result = Vec::with_capacity(count);
+    for i in 0..count {
+        result.push(create_row_key(1, i as i64));
```
Comment (zhangjinpeng1987) [inline benches/bin/mvcctxn.rs]: It's better to make table_id variable, and make handle_id as a variable range.

Comment (zhangjinpeng1987) [reply benches/bin/mvcctxn.rs]: BTW, index key also need test.

**Comment by zhangjinpeng1987** [comment]
Batch insert testing should included. BTW, any conclusion with this bench?

File: benches/bin/mvcctxn.rs (hunk 9 above)
Comment (zhangjinpeng1987) [reply benches/bin/mvcctxn.rs]: `fn generate_keys(start_id: u64, count: usize) -> Vec<Vec<u8>>`

File: benches/bin/bench-tikv.rs [hunk 10]
```diff
@@ -96,6 +103,44 @@ fn main() {
         );
     }
     // TODO allow user to specify flag to just bench some cases.
-    raftstore::bench_raftstore();
-    mvcc::bench_engine();
+
```
Comment (zhangjinpeng1987) [inline benches/bin/bench-tikv.rs]: Please remove above TODO.

File: src/storage/engine/mod.rs [hunk 11]
```diff
@@ -520,21 +520,31 @@ impl Cursor {
...
             _ => CFOptions::new(*cf, ColumnFamilyOptions::new()),
         };
         cfs_opts.push(cf_opt);
     }
-    EngineRocksdb::new(path, cfs, Some(cfs_opts)).map(|engine| -> Box<Engine> { Box::new(engine) })
+    EngineRocksdb::new(path, cfs, Some(cfs_opts), with_wal)
+        .map(|engine| -> Box<Engine> { Box::new(engine) })
+}
+
+/// Create a local Rocskdb engine with default config. (Without raft, mainly for tests).
```
Comment (zhangjinpeng1987) [inline src/storage/engine/mod.rs]: The functional comment is redundant with above functional comment.

File: benches/bin/utils.rs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+/// Convert duration to nanoseconds
+pub fn to_total_nanos(duration: &Duration) -> u64 {
+    duration.as_secs() * 1_000_000_000 + (duration.subsec_nanos() as u64)
+}
+
+/// Run `job` for `iterations` times, and return a Vec containing nanoseconds of each turn `job`
+/// returns.
+///
+/// `job` must return a `Duration` type. Please calculate time cost of what you want to bench
+/// manually ant return it.
```
Comment (zhangjinpeng1987) [inline benches/bin/utils.rs]: s/ant/and

File: benches/bin/utils.rs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+///
+/// Attention that too short job may cause great inaccuracy.
+pub fn record_time<F>(mut job: F, iterations: u32) -> Vec<u64>
```
Comment (zhangjinpeng1987) [inline benches/bin/utils.rs]: It's better to calculate time once after all jobs finished, not for each job once.

File: benches/bin/mvcctxn.rs [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+    let mut txn = MvccTxn::new(snapshot, start_ts, None, IsolationLevel::SI, false);
+    for key in keys {
+        txn.commit(key, commit_ts).unwrap();
+    }
+    do_write(engine, txn.into_modifies());
+}
+
+fn prepare_test_engine(versions: usize, value_len: usize, keys: &[Vec<u8>]) -> Box<Engine> {
+    let mut config = DbConfig::default();
+    config.defaultcf.write_buffer_size = ReadableSize::gb(1);
```
Comment (zhangjinpeng1987) [inline benches/bin/mvcctxn.rs]: Please give an explanation why write buffer size is 1GB.

File: benches/bin/mvcctxn.rs [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+
+#[inline]
+fn get(engine: &Engine, key: &Key, statistics: &mut Statistics) -> Option<Vec<u8>> {
+    let snapshot = engine.snapshot(&Context::new()).unwrap();
+    let start_ts = next_ts();
+    let snapstore = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, false);
+    snapstore.get(key, statistics).unwrap()
+}
+
+fn bench_get(engine: &Engine, keys: &[Vec<u8>]) -> BenchSamples {
```
Comment (zhangjinpeng1987) [inline benches/bin/mvcctxn.rs]: For get, we should use an appropriate block-cache-size and turn `fill_cache` on.

**Comment by zhangjinpeng1987** [comment]
Good job.

**Comment by zhangjinpeng1987** [comment]
It's better also to output QPS.

File: benches/bin/mvcctxn.rs [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+        (ns / (batch_size as f64)) as u64,
+        (1_000_000_000_f64 * (batch_size as f64) / ns) as u64
+    );
+}
+
+
+
+pub fn bench_mvcctxn() {
+    for bench_type in &[BenchType::Row, BenchType::UniqueIndex] {
+        // for table_size in &[1_000, 10_000, 100_000] {
```
Comment (zhangjinpeng1987) [inline benches/bin/mvcctxn.rs]: Please remove these lines.

File: benches/bin/mvcctxn.rs [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+}
+
+fn bench_delete(db: Arc<DB>, keys: &[Vec<u8>]) -> f64 {
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            let key = &keys[rng.gen_range(0, keys.len())];
```
Comment (zhangjinpeng1987) [inline benches/bin/mvcctxn.rs]: Some rows may have been already deleted?

Comment (MyonKeminta) [reply benches/bin/mvcctxn.rs]: I didn't find any logic  in `MvccTxn` that handles deleting a already-deleted key, but simply insert a delete record in `CF_WRITE`. So I think it's ok to do this.

File: benches/bin/mvcctxn.rs [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+use utils::*;
+
+
+
+#[inline]
+fn do_write(db: &DB, modifies: Vec<Modify>) {
+    let wb = WriteBatch::new();
+    for rev in modifies {
+        match rev {
+            Modify::Delete(cf, k) => if cf == CF_DEFAULT {
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: You can get `CFHandle` of `CF_DEFAULT` too, so it is not necessary to distinguish them.

File: benches/bin/mvcctxn.rs [hunk 19]
```diff
@@ -0,0 +1,472 @@
...
+    db.write_without_wal(wb).unwrap();
+}
+
+#[inline]
+fn get_snapshot(db: Arc<DB>) -> Box<Snapshot> {
+    box SyncSnapshot::new(db) as Box<Snapshot>
+}
+
+#[inline]
+fn prewrite(db: Arc<DB>, mutations: &[Mutation], primary: &[u8], start_ts: u64) {
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: Is it better to collect these db/txn related wrappers `get/write/prewrite/commit` into a struct?

File: benches/bin/mvcctxn.rs [hunk 20]
```diff
@@ -0,0 +1,472 @@
...
+}
+
+fn prepare_test_db(versions: usize, value_len: usize, keys: &[Vec<u8>], path: &str) -> Arc<DB> {
+    let mut config = DbConfig::default();
+    // Use a huge write_buffer_size to avoid flushing data to disk.
+    config.defaultcf.write_buffer_size = ReadableSize::gb(1);
+    config.writecf.write_buffer_size = ReadableSize::gb(1);
+    config.lockcf.write_buffer_size = ReadableSize::gb(1);
+
+    let cf_ops = vec![
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: `config.build_cf_opts()`?

File: benches/bin/mvcctxn.rs [hunk 21]
```diff
@@ -0,0 +1,472 @@
...
+    let snapstore = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
+    snapstore.get(key, statistics).unwrap()
+}
+
+fn bench_get(db: Arc<DB>, keys: &[Vec<u8>]) -> f64 {
+    let mut fake_statistics = Statistics::default();
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let index = rng.gen_range(0, keys.len());
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: `rng.choose(keys)`

File: benches/bin/mvcctxn.rs [hunk 22]
```diff
@@ -0,0 +1,472 @@
...
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            keys_to_write.clear();
+            mutations.clear();
+
+            // Randomly select non-duplicated keys
+            for i in 0..batch_size {
+                let selected = rng.gen_range(i, keys.len());
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: Maybe you can shuffle keys before the bench.

Comment (MyonKeminta) [reply benches/bin/mvcctxn.rs]: here I want to select some distinct keys randomly from `keys` in each iteration...

File: benches/bin/mvcctxn.rs (hunk 19 above)
Comment (zhangjinpeng1987) [reply benches/bin/mvcctxn.rs]: ？

**Comment by zhangjinpeng1987** [comment]
LGTM

File: benches/bin/bench-tikv.rs [hunk 23]
```diff
@@ -95,7 +102,44 @@ fn main() {
             e
         );
     }
-    // TODO allow user to specify flag to just bench some cases.
-    raftstore::bench_raftstore();
-    mvcc::bench_engine();
+
+
+    let mut args: Vec<_> = env::args().skip(1).collect();
```
Comment (siddontang) [inline benches/bin/bench-tikv.rs]: please use another PR for it.

Comment (MyonKeminta) [reply benches/bin/bench-tikv.rs]: #2637

File: benches/bin/bench-tikv.rs [hunk 24]
```diff
@@ -101,7 +102,7 @@ fn main() {
         );
     }
 
-    let available_benches = ["raftstore", "mvcc"];
+    let available_benches = ["raftstore", "mvcc", "mvcctxn"];
```
Comment (siddontang) [inline benches/bin/bench-tikv.rs]: naming txn is enough.

File: benches/bin/mvcctxn.rs [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+    table_size: usize,
+    version_count: usize,
+    data_len: usize,
+    bench_type: &BenchType,
+) {
+    let (mut keys, value_len, log_name) = match *bench_type {
+        BenchType::Row => (generate_row_keys(1, 0, table_size), data_len, "row"),
+        BenchType::UniqueIndex => (
+            generate_unique_index_keys(1, 1, data_len, table_size),
+            8,
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: Why 8 here?

Comment (MyonKeminta) [reply benches/bin/mvcctxn.rs]: In a real unique index, the value of the key-value pair stores RowID which is an int64

File: benches/bin/bench-tikv.rs [hunk 26]
```diff
@@ -127,6 +128,7 @@ fn main() {
         match item {
             "raftstore" => raftstore::bench_raftstore(),
             "mvcc" => mvcc::bench_engine(),
+            "txn" => mvcctxn::bench_mvcctxn(),
```
Comment (huachaohuang) [inline benches/bin/bench-tikv.rs]: I guess @siddontang means to replace all "mvcctxn" to "txn".

File: benches/bin/mvcctxn.rs [hunk 27]
```diff
@@ -0,0 +1,248 @@
...
+        &keys,
+        dir.path().to_str().unwrap(),
+    );
+
+    println!(
+        "benching mvcctxn {} get\trows:{} versions:{} data len:{}\t...",
+        log_name, table_size, version_count, data_len
+    );
+    let ns = bench_get(Arc::clone(&db), &keys) as u64;
+    println!("\t{:>11} ns per op  {:>11} ops", ns, 1_000_000_000 / ns);
```
Comment (huachaohuang) [inline benches/bin/mvcctxn.rs]: Can you print the time in a more friendly format?

**Comment by MyonKeminta** [comment]
@zhangjinpeng1987 PTAL

**Comment by huachaohuang** [comment]
How we do with this PR now?

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Pull request closed without merging**